### PR TITLE
Support collections for routing members (breaking change)

### PR DESCRIPTION
**Pull request by ocoanet**
The goal of this PR is to support the following syntax for dynamic subscriptions: 
```cs
bus.Subscribe(Subscription.Matching<Message>(x => x.Ids.Contains(999));
```

In order to make this change as transparent as possible and to avoid compatibility issues:
- The "contains" subscriptions will have a standard format similar to "equals" subscriptions
- The only impact should be on the publishing side where the subscription is evaluated against the collection.

This PR also introduces the RoutingContent type, which stores the routing member values of routable messages. Before, BindingKey was used both to store subscription routing keys and routing member values. This change should make the types usages more explicit.

### CODE REVIEW

File: src/Abc.Zebus/Routing/RoutingContentValue.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    /// <summary>
+    /// Stores the routing member value of a routable message.
+    /// </summary>
+    [StructLayout(LayoutKind.Explicit)]
+    public readonly struct RoutingContentValue : IEquatable<RoutingContentValue>
+    {
+        [FieldOffset(0)]
+        private readonly string? _value;
+        [FieldOffset(0)]
+        private readonly string?[] _values;
```
[the remark this answers is not in the document]
Comment (ocoanet) [reply src/Abc.Zebus/Routing/RoutingContentValue.cs:17]: Well, I do not want the value to be nullable when `_isCollection` is true.

File: src/Abc.Zebus/Routing/RoutingContentValue.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        private readonly bool _isCollection;
+
+        public RoutingContentValue(string? value)
+        {
+            _values = null!;
+            _value = value;
+            _isCollection = false;
+        }
+
+        public RoutingContentValue(string?[] values)
```
[the remark this answers is not in the document]
Comment (ocoanet) [reply src/Abc.Zebus/Routing/RoutingContentValue.cs:28]: Possibly, why?

File: src/Abc.Zebus/Directory/MessageBinding.cs [hunk 3]
```diff
@@ -3,36 +3,42 @@
 
 namespace Abc.Zebus.Directory
 {
-    public readonly struct MessageBinding : IEquatable<MessageBinding>
```
[the remark this answers is not in the document]
Comment (ocoanet) [reply src/Abc.Zebus/Directory/MessageBinding.cs:6]: Some types are designed to be equatable or comparable (ID or key types). I do not like to implement `IEquatable` without a reason.

File: src/Abc.Zebus/MessageTypeDescriptor.cs [hunk 4]
```diff
@@ -68,81 +69,111 @@ internal static MessageTypeDescriptor Load(Type? messageType, string? fullName)
...
-            private static MethodCallExpression BuildToStringExpression(Expression castExpression, MemberInfo member)
+            private static RoutingContentValue GetMemberValue<TMember>(TMember? value)
+                => new RoutingContentValue(value?.ToString());
+
+            private static RoutingContentValue GetMemberValueConvertible<TMember>(TMember? value)
+                where TMember : IConvertible
+                => new RoutingContentValue(value?.ToString(CultureInfo.InvariantCulture));
+
+            private static RoutingContentValue GetMemberValues<TMember>(ICollection<TMember>? values)
+                => new RoutingContentValue(values != null ? values.Select(x => x!.ToString()).ToArray() : Array.Empty<string>());
```
[the remark this answers is not in the document]
Comment (ocoanet) [reply src/Abc.Zebus/MessageTypeDescriptor.cs:129]: I will change that to support the evil case of null values inside collections :)

File: src/Abc.Zebus/MessageTypeDescriptor.cs [hunk 5]
```diff
@@ -68,81 +69,111 @@ internal static MessageTypeDescriptor Load(Type? messageType, string? fullName)
...
                 {
-                    var fieldInfo = (FieldInfo)member;
-                    memberAccessor = m => Expression.Field(m, fieldInfo);
-                    memberType = fieldInfo.FieldType;
+                    var castExpression = Expression.Convert(ParameterExpression, member.DeclaringType!);
+
+                    return member switch
+                    {
+                        PropertyInfo propertyInfo => (Expression.Property(castExpression, propertyInfo), propertyInfo.PropertyType),
+                        FieldInfo fieldInfo       => (Expression.Field(castExpression, fieldInfo), fieldInfo.FieldType),
```
[the remark this answers is not in the document]
Comment (ocoanet) [reply src/Abc.Zebus/MessageTypeDescriptor.cs:149]: `PropertyOrField` is going to load the member from the name but I can probably use `Expression.MakeMemberAccess`

File: src/Abc.Zebus.Tests/Directory/PeerSubscriptionTreeTests.cs [hunk 6]
```diff
@@ -264,10 +265,42 @@ public void invalid_subscription_should_not_match_routing_key(string subscriptio
...
             // Act
-            var matchingPeers = peerSubscriptionTree.GetPeers(BindingKeyHelper.CreateFromString(routingKey, '.'));
+            var routingContent = RoutingContent.FromValues(routingKey.Split('.'));
+            var matchingPeers = peerSubscriptionTree.GetPeers(routingContent);
 
             // Assert
             matchingPeers.ShouldBeEmpty();
         }
+
+        // TODO: investigate this test case
```
[the remark this answers is not in the document]
Comment (ocoanet) [reply src/Abc.Zebus.Tests/Directory/PeerSubscriptionTreeTests.cs:275]: I wonder if we want to support this for compatibility reasons (adding new routing members).

File: src/Abc.Zebus/Subscription.cs [hunk 7]
```diff
@@ -166,27 +171,46 @@ public static Subscription Matching<TMessage>(Expression<Func<TMessage, bool>> p
...
+                    AddFieldValueFromMemberExpression<TMessage>(fieldValues, memberExpression);
+                    return;
+
+                case MethodCallExpression methodCallExpression when methodCallExpression.Method.Name == "Contains":
+                    if (methodCallExpression.Object != null && methodCallExpression.Arguments.Count == 1)
+                    {
+                        AddFieldValueFromContainsExpression<TMessage>(fieldValues, methodCallExpression, methodCallExpression.Object, methodCallExpression.Arguments[0]);
+                        return;
+                    }
+                    if (methodCallExpression.Object == null && methodCallExpression.Arguments.Count == 2)
```
[the remark this answers is not in the document]
Comment (ocoanet) [reply src/Abc.Zebus/Subscription.cs:194]: `methodCallExpression.Object == null` implies that the method is static

File: src/Abc.Zebus/Routing/RoutingContentValue.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ltrzesniewski) [reply src/Abc.Zebus/Routing/RoutingContentValue.cs:17]: Fair enough, that invariant is kept with `default(RoutingContentValue)`.

**Pull request merged**